src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -151,11 +172,35 @@ async def collect_thirdparty_modules(
         for target in applicable_lockfile_entries.values()
     )
 
-    resolve_digest = await Get(Digest, MergeDigests(cpe.digest for cpe in classpath_entries))
+    digests = []
+    for cpe in classpath_entries:
+        digests.append(cpe.digest)
+    for alse in applicable_lockfile_source_entries:
+        new_file = FileEntry(alse.file_name, alse.file_digest)
+        digest = await Get(Digest, CreateDigest([new_file]))
+        digests.append(digest)
+
+        for dep in alse.dependencies:
+            coord = Coordinate.from_coord_str(dep.to_coord_str())
+            dep_artifact_requirement = ArtifactRequirement(coord)
+            dep_entry = get_entry_for_coord(lockfile, dep_artifact_requirement.coordinate)
+            dep_new_file = FileEntry(dep_entry.file_name, dep_entry.file_digest)
+            dep_digest = await Get(Digest, CreateDigest([dep_new_file]))
+            digests.append(dep_digest)
+            src_ent = applicable_lockfile_source_entries.get(alse)
+            applicable_lockfile_source_entries_inverse.get(src_ent).append(dep_entry)
+
+    resolve_digest = await Get(Digest, MergeDigests(digests))
+    inverse = dict(zip(classpath_entries, applicable_lockfile_entries))
+
+    s = map(
+        lambda x: (x, applicable_lockfile_source_entries_inverse.get(inverse.get(x))),
+        classpath_entries,
+    )
 
     return ThirdpartyModules(
         resolve,
-        dict(zip(applicable_lockfile_entries, classpath_entries)),
+        dict(zip(applicable_lockfile_entries, s)),
```

Comment:
Convert `s` to tuple here given earlier comment to not include `list` in a frozen dataclass.